kernel/fs.c

```diff
@@ -109,42 +112,65 @@ int myst_load_fs(
             if (locals->fssig.signature_size > sizeof(locals->fssig.signature))
                 ERAISE(-EINVAL);
 
-            ECHECK(myst_pubkey_verify(
+            pubkey_verify_ret = myst_pubkey_verify(
                 __myst_kernel_args.pubkeys_data,
                 __myst_kernel_args.pubkeys_size,
                 locals->fssig.root_hash,
                 sizeof(locals->fssig.root_hash),
                 locals->fssig.signer,
                 sizeof(locals->fssig.signer),
                 locals->fssig.signature,
-                locals->fssig.signature_size));
+                locals->fssig.signature_size);
         }
-        else
+
+        /* If public key verification failed, try roothash verification */
+        if (pubkey_verify_ret != 0)
         {
-            ECHECK(myst_roothash_verify(
+            roothash_verify_ret = myst_roothash_verify(
                 __myst_kernel_args.roothashes_data,
                 __myst_kernel_args.roothashes_size,
                 locals->fssig.root_hash,
-                sizeof(locals->fssig.root_hash)));
+                sizeof(locals->fssig.root_hash));
+
+            if (roothash_verify_ret == 0)
+                pubkey_verify_ret = 0;
         }
 
-        /* create the device stack */
-        ECHECK(myst_verityblkdev_open(
-            source,
-            locals->fssig.hash_offset,
-            locals->fssig.root_hash,
-            sizeof(myst_sha256_t),
-            &blkdev));
+        /* if the roothash was verified */
+        if (roothash_verify_ret == 0 && pubkey_verify_ret == 0)
+        {
+            /* create the device stack */
+            ECHECK(myst_verityblkdev_open(
+                source,
+                locals->fssig.hash_offset,
+                locals->fssig.root_hash,
+                sizeof(myst_sha256_t),
+                &blkdev));
+        }
     }
     else
     {
-#ifdef MYST_RELEASE
-        /* Fail if neither a roothash nor a public key were provided */
-        ERAISE(-EPERM);
-#else
-        const bool ephemeral = true;
-        ECHECK(myst_rawblkdev_open(source, ephemeral, 0, &blkdev));
-#endif
+        pubkey_verify_ret = -EACCES;
+        roothash_verify_ret = -EACCES;
+    }
+
+    /* if failed to verify roothash (using one method or the other) */
+    if (roothash_verify_ret != 0 || pubkey_verify_ret != 0)
+    {
+        /* if in debug mode, allow mount in spite of errors */
+        if (__myst_kernel_args.tee_debug_mode)
+        {
+            const bool ephemeral = true;
```

Comment:
I replaced both variables with a single boolean called ``verify``. Either verification by pubkey works (setting ``verify`` to true) or verification by roothash works (setting ``verify`` to true). If ``verify`` is ultimately false, then we allow if debug is enabled. There is no **present and rejected** because verification only requires a roothash or a pubkey that verifies the EXT2 image. Since there can be multiple of each, we only care to know that no mechanism was able to verify.

Reply:
In other words, the **roothash list** and the **pubkey list** function as "allow lists" not as "deny lists".

Reply:
Because then all of our testing would be targeted a the non-production block device (rawblkdev) rather than the production block device (verityblkdev).